### ISSUE TITLE
DNS resolution not respecting search domain

### ISSUE THREAD

**Issue by thomseno** [docker/for-mac#1763]
### Expected behavior

Inside a Debian based container resolving hostnames should respect the search domain (mydomain.com):

"ping hostname" should give the same result as "ping hostname.mydomain.com"

### Actual behavior

- ping hostname results in "ping: unknown host"
- ping hostname.mydomain.com is working

### Information

Diagnostic-ID: 9E3B454E-2BF2-4FC3-B27D-327B5B5C90C3
Docker for Mac: version: 17.06.0-ce-rc5-mac16 (4cdec4294a50b2233146b09469b49937dabdebdd)
macOS: version 10.12.5 (build: 16F73)
logs: /tmp/9E3B454E-2BF2-4FC3-B27D-327B5B5C90C3/20170622-164003.tar.gz
[OK]     db.git
[OK]     vmnetd
[OK]     dns
[OK]     driver.amd64-linux
[OK]     virtualization VT-X
[OK]     app
[OK]     moby
[OK]     system
[OK]     moby-syslog
[OK]     db
[OK]     env
[OK]     virtualization kern.hv_support
[OK]     slirp
[OK]     osxfs
[OK]     moby-console
[ERROR]  logs
         logs check failed with: (Failure
  "exec: /usr/bin/log show --style syslog --last 4d | grep -i docker > \"/tmp/9E3B454E-2BF2-4FC3-B27D-327B5B5C90C3/20170622-164003/docker-system-os.log\": exit 1")
[OK]     docker-cli
[OK]     menubar
[OK]     disk

### Steps to reproduce the behavior

  1. docker run --rm -it buildpack-deps:jessie-scm /bin/bash
  2. ping a-host-in-local-network-without-specifying-the-domain

### Further information

- Container based on "alpine" or "busybox" are working.
- Container based on "centos" also not working.
- Docker for Mac 17.03.1-ce is also working

**Comment by kinghuang** [comment]
In step 1, you haven't specified any search domains. If you want to include `mydomain.com` in the search domains, it can be set using the `--dns-search` option.

```
docker run --rm -it --dns-search=mydomain.com buildpack-deps:jessie-scm /bin/bash
```

**Comment by thomseno** [comment]
Good point...

My assumption was, that Docker would pick up the settings from host network settings. But why is it working for some containers and for others not?

**Comment by kinghuang** [comment]
Not sure. Maybe it's possible with `--net=host`?

**Comment by thomseno** [comment]
But it is working with current version of Docker for Mac.

I've discovered the behavior when updating Docker to edge version (was hoping for speedup from new volume caching), because Jenkins container was not finding local servers for deployment... before the update it was working perfectly.

**Comment by kinghuang** [comment]
Hmm, I don't see anything obvious in the [release notes](https://docs.docker.com/docker-for-mac/release-notes/#docker-community-edition-17060-rc1-ce-mac13-2017-06-01-edge) about changes related to search domains. Hopefully someone from Docker will comment.

**Comment by FelikZ** [comment]
I have the same issue. Just migrated from stable 17.03 to 17.06 and it does not resolve, so sad.

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 17.06.0-ce-rc5
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.31-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.787GiB
Name: moby
ID: 4LZG:GZX3:CFS6:DJP5:C26Y:UWSL:L7Y2:WQVT:RZ5Q:QOCC:UTG7:7462
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 17
 Goroutines: 28
 System Time: 2017-06-23T16:13:41.39313636Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 docker-reg.textkernel.local:5000
 127.0.0.0/8
Live Restore Enabled: false
```

Workaround is to put network domain name into `/etc/hosts/, no docker restart required.

**Comment by djs55** [comment]
I think there may be a problem with the DHCP response to the VM failing to propagate the host's search domain to the VM.

In System Preferences -> Network -> Advanced -> DNS my Search domain is `docker.com`.

On `Version 17.03.1-ce-mac12 (17661) Channel: stable d1db12684b`
```
$ docker run --rm --pid=host --privileged -it justincormack/nsenter1 /bin/cat /etc/resolv.conf
# Generated by dhcpcd from eth0.dhcp
# /etc/resolv.conf.head can replace this line
domain docker.com
nameserver 192.168.65.1
# /etc/resolv.conf.tail can replace this line
```

Whereas on `Version 17.06.0-ce-rc5-mac16 (18400) Channel: edge 6cc0236e3b`
```
$ docker run --rm --pid=host --privileged -it justincormack/nsenter1 /bin/cat /etc/resolv.conf
# Generated by dhcpcd from eth0.dhcp
# /etc/resolv.conf.head can replace this line
domain local
nameserver 192.168.65.1
# /etc/resolv.conf.tail can replace this line
```

I'd be interested to know if manually logging into the VM with `docker run --rm --net=host --pid=host --privileged -it justincormack/nsenter1 /bin/sh` and fixing the `domain` in the `/etc/resolv.conf` has any effect or not. I'll investigate fixing this particular problem, but it's possible there's more than one bug involved.

Thanks for your report!

**Comment by djs55** [comment]
For testing purposes only, there is a build of Docker for Mac with the fix for the missing search domain bug:

https://download-stage.docker.com/mac/master/18442/Docker.dmg

If you install this build then on a host with search domain `foo.com` and domain name `bar.com`, running

```
docker run --rm --net=host --pid=host --privileged -it justincormack/nsenter1 cat /etc/resolv.conf
```

should show something like
```
# Generated by dhcpcd from eth0.dhcp
# /etc/resolv.conf.head can replace this line
domain bar.com
search foo.com
nameserver 192.168.65.1
# /etc/resolv.conf.tail can replace this line
```

If you have the time to try this, let me know the results. Thanks again for your report!

**Comment by thomseno** [comment]
Downloaded the provided version, installed and tested it: looks good :+1:

Finally I can try the new volume caching parameters, thanks!

**Comment by thomseno** [comment]
Downloaded, installed & tested: looks good 👍 

Tried it also on my Jenkins container and everything works as expected.

Thanks for the fix!

**Comment by djs55** [comment]
Thanks for the quick confirmation!

Unfortunately this particular fix has missed the deadline for the imminent 17.06 stable release. It will be in the next 17.07 edge release which should be out in a few weeks.

**Comment by thomseno** [comment]
Thanks... once again... so sad the fix didn't make it into the 17.06 release 😞 

@djs Will you close the issue?

**Comment by djs55** [comment]
I'll close it. Thanks again for your report!
